File: dailyChallenges/0016-WordPattern/16-WordPattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        word = s.split(" ");
        charToWords, wordsToChar = {}, {}

        if len(pattern) != len(word):
            return False

        for c, w in zip(pattern, word):
            if c in charToWords and charToWords[c] != w:
                return False
            if w in wordsToChar and wordsToChar[w] != c:
                return False
            charToWords[c] = w
            wordsToChar[w] = c

        return True

File: dailyChallenges/0016-WordPattern/16-WordPattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        std::vector<std::string> words;
        std::istringstream iss(s);
        std::string word;
        std::unordered_map<char, std::string> charToWords;
        std::unordered_map<std::string, char> wordsToChar;

        while(iss >> word)
        {
            words.push_back(word);
        }

        if(pattern.length() != words.size())
        {
            return false;
        }

        for (int i=0; i<pattern.length(); i++)
        {
            char c = pattern[i];
            std::string w = words[i];

            if((charToWords.find(c) != charToWords.end()) && (charToWords[c] != w))
            {
                return false;
            }
            if((wordsToChar.find(w) != wordsToChar.end()) && (wordsToChar[w] != c))
            {
                return false;
            }
            charToWords[c] = w;
            wordsToChar[w] = c;
        }
    return true;
    }
};